### ISSUE TITLE
Fault-tolerant checklist

### ISSUE THREAD

**Issue by awaelchli** [PyTorchLightning/pytorch-lightning#9130]
## 🚀 Fault-tolerant training progress tracker

Fault-tolerant training in PL can be activated by setting the environment variable. 

```
PL_FAULT_TOLERANT_TRAINING=1
```

[Description of the architecture around fault-tolerant training
](https://docs.google.com/document/d/1pvhyNg17pm7-ECMTpgpZiJf6sbypeuaB8QsSP79n6k0/edit?usp=sharing)

- [x] progress tracking in loops
- [x] progress restart of loops
- [ ] capturing state of iterable datasets
   - [x] non-random state, multiple workers
   - [ ] random state
   - [x] sampler inside of iterable dataset
   - [ ] batch sampler
   - [ ] testing of items above
   
- [ ] capturing state of map-style dataset
   - [x] single worker global random state (#8950)
   - [x] multiple workers global random state (#8950)
   - [ ] local generators
   - [ ] random sampler state, after updates in [pytorch 1.10](https://github.com/pytorch/pytorch/pull/63026)
   - [ ] testing of items above (partially done in #8950)

- [ ] mid epoch restart
   - [ ] failure on mid training epoch loop
      - [x] failure on mid training epoch loop with 1 optimizer
      - [x] failure on mid training epoch loop with multiple optimizers  #9537 
      - [ ] failure between optimizer step and scheduler step

   - [x] failure on mid validation epoch loop: #9563, #9681
      - [x] reset on epoch end validation with 1 dataloaders
      - [x] reset on epoch end validation with multiple dataloaders
      - [ ] reset on validation using val_check_interval: https://github.com/PyTorchLightning/pytorch-lightning/pull/9563/files#diff-c2e3dba8e80f865d593f0310cdf80eaeccdb80c9a250f7eda9edd466b7710874R989-R995

- [ ] epoch end restart
   - [ ] failure on training epoch loop end
      - [ ] failure between optimizer step and scheduler step
       - [ ] failure before logging metrics
   - [ ] no failure on training epoch end

- [ ] reloading multiple train dataloaders with individual random states (see #8950 for todos in added tests)
- [ ] Fault-tolerant training with multiple dataloaders (#11349)
- [x] parametrize the important tests in `test_auto_restart.py` with num_workers > 0 and run as part of slow tests #9086  
- [x] end-user guide - how to use and description of limitations  #9256 
- [ ] benchmark iteration/s for small and large models with fault-tolerant enabled vs. not enabled
- [x] add test case for different batch structures (dict, list, etc.)  #9281 
- [x] fix progress bar tracking on restart #9310
- [ ] Add logic to resume OptimizerLoop? Is this needed/wanted?
- [ ] Fix num_workers > 0 causing repeated random state when resuming
- [ ] Fix TODOs
    https://github.com/PyTorchLightning/pytorch-lightning/blob/ce00053002a1bb5385f7e44fceea97af50313d4c/tests/loops/test_loops.py#L899-L900
    https://github.com/PyTorchLightning/pytorch-lightning/blob/5841ca97825bd9786ab84d70f0abfa6e673528b4/tests/trainer/connectors/test_signal_connector.py#L27
    https://github.com/PyTorchLightning/pytorch-lightning/blob/5841ca97825bd9786ab84d70f0abfa6e673528b4/tests/utilities/test_auto_restart.py#L1181-L1183

Open questions:
- Do we need to capture `torch.cuda.get_rng_state` too?

Supplementary documents:
[State of fault-tolerant training in PyTorch Lightning](https://docs.google.com/document/d/1pvhyNg17pm7-ECMTpgpZiJf6sbypeuaB8QsSP79n6k0/edit?usp=sharing)

cc @borda @carmocca @justusschock @awaelchli @ninginthecloud

**Comment by aazzolini** [comment]
Do we have a design document for the proposed work?
Fault tolerance is very easy to get wrong. We have spent a lot of time at Facebook ensuring fault tolerance in our internal toolkit so we have some idea of multiple things that could go wrong.
It would be great if we went through a careful design review for this one.

**Comment by ananthsub** [comment]
@awaelchli @tchaton  - I don't think metrics is properly covered here. One gap today is around the assumptions o saving the metric states only on rank 0. https://github.com/PyTorchLightning/pytorch-lightning/blob/35876bb75f27eb8f44220afd4bfa757a0432d233/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L137-L141 

Doesn't this not hold for deepspeed, or any case where we save a part of the lightning module state dict across ranks?

**Comment by tchaton** [comment]
> Do we have a design document for the proposed work?
> Fault tolerance is very easy to get wrong. We have spent a lot of time at Facebook ensuring fault tolerance in our internal toolkit so we have some idea of multiple things that could go wrong.
> It would be great if we went through a careful design review for this one.

Hey @aazzolini,

I entirely agree, this is definitely very error prone ! We are working on this with @awaelchli and we will share it with you asap.
We would be happy to have recorded meeting to present the code too for ease onboarding.
We will really appreciate if Facebook could participate on this feature support, as I believe PyTorch needs improvement to make this process simpler.

Hey @ananthsub,

We didn't investigate the impact of sharding on TorchMetrics yet and added fault tolerant for non-sharded model for our fault tolerant V0. 
IMO, we should have a mechanism to prevent sharding by default for Metric as each rank is accumulating their own state.   
 
My intuition is that this assumption should still hold. Before saving, we accumulate the states on all ranks. So each ranks will contain the accumulated states. On reload, we just reset the metric on non-rank 0. 

But it might become more intricate if metrics states get sharded, which might results in state collisions.

Best,
T.C
